src/Compilers/Core/Portable/PEWriter/DebugDirectoryExtensions.cs

```diff
@@ -0,0 +1,123 @@
+        {
+            int start = builder.Count;
+
+            builder.WriteByte((byte)'P');
+            builder.WriteByte((byte)'D');
+            builder.WriteByte((byte)'B');
+            builder.WriteByte((byte)'C');
+
+            // NUL-terminated algorithm name:
+            builder.WriteUTF8(algorithmName, allowUnpairedSurrogates: true);
+            builder.WriteByte(0);
+
+            // checksum:
+            builder.WriteBytes(checksum);
+
+            return builder.Count - start;
+        }
+
+        /// <summary>
+        /// Reads the data pointed to by the specified Debug Directory entry and interprets them as PDB Checksum entry.
+        /// </summary>
+        /// <exception cref="ArgumentException"><paramref name="entry"/> is not a CodeView entry.</exception>
+        /// <exception cref="BadImageFormatException">Bad format of the data.</exception>
+        /// <exception cref="IOException">IO error while reading from the underlying stream.</exception>
+        /// <exception cref="InvalidOperationException">PE image not available.</exception>
+        internal static PdbChecksumDebugDirectoryData ReadPdbChecksumDebugDirectoryData(this PEReader peReader, DebugDirectoryEntry entry)
+        {
+            if (entry.Type != PdbChecksumEntryType)
+            {
+                throw new ArgumentException("Unexpected debug directory entry type", nameof(entry));
+            }
+
+            var peImage = peReader.GetEntireImage();
+
+            int dataOffset = peReader.IsLoadedImage ? entry.DataRelativeVirtualAddress : entry.DataPointer;
+            var reader = peImage.GetReader(dataOffset, entry.DataSize);
+
+            if (reader.ReadByte() != (byte)'P' ||
+                reader.ReadByte() != (byte)'D' ||
+                reader.ReadByte() != (byte)'B' ||
+                reader.ReadByte() != (byte)'C')
+            {
+                throw new BadImageFormatException("Unexpected PDB Checksum signature");
+            }
+
+            int nameLength = reader.IndexOf(0);
+            if (nameLength <= 0)
+            {
+                throw new BadImageFormatException("Invalid PDB Checksum data format");
+            }
+
+            string algorithmName = reader.ReadUTF8(nameLength);
+
+            // NUL
+            reader.Offset += 1;
+
+            var checksum = reader.ReadBytes(reader.RemainingBytes).ToImmutableArray();
+            if (checksum.Length == 0)
+            {
+                throw new BadImageFormatException("Invalid PDB Checksum data format");
+            }
+
+            return new PdbChecksumDebugDirectoryData(algorithmName, checksum);
+        }
+    }
+
+    internal readonly struct PdbChecksumDebugDirectoryData
+    {
+        /// <summary>
+        /// Checksum algorithm name.
+        /// </summary>
+        public string AlgorithmName { get; }
+
+        /// <summary>
+        /// GUID (Globally Unique Identifier) of the associated PDB.
```

Comment:
>GUID (Globally Unique Identifier) of the associated PDB. [](start = 12, length = 56)

Maybe improve the comment. I think it is a checksum and we could expect it to be unique but it maybe uncommon to refer it as GUID.

Reply:
Yes, the comment is wrong.